File: ElasticSearch/TS/complete.ts
import { Client } from '@elastic/elasticsearch'

const client = new Client({ node: 'http://localhost:9200' })

// Utility function to log results
const logResult = (operation: string, result: any) => {
  console.log(`${operation}:`, result)
}

// Create: Index Document
const createDocument = async () => {
  const doc = { field: 'value' }
  const res = await client.index({
    index: 'index',
    id: 'id',
    body: doc,
    refresh: 'true',
  })
  logResult('createDocument', res)
}

// Create: Bulk Index Documents
const bulkIndexDocuments = async () => {
  const bulk = [
    { index: { _id: '1' } },
    { field: 'value1' },
    { index: { _id: '2' } },
    { field: 'value2' },
  ]
  const res = await client.bulk({
    index: 'index',
    body: bulk,
    refresh: 'true',
  })
  logResult('bulkIndexDocuments', res)
}

// Read: Get Document
const getDocument = async () => {
  const res = await client.get({
    index: 'index',
    id: 'id',
  })
  logResult('getDocument', res)
}

// Read: Search Documents
const searchDocuments = async () => {
  const query = {
    query: {
      match: {
        field: 'value',
      },
    },
  }
  const res = await client.search({
    index: 'index',
    body: query,
    pretty: true,
  })
  logResult('searchDocuments', res)
}

// Read: Count Documents
const countDocuments = async () => {
  const query = {
    query: {
      match_all: {},
    },
  }
  const res = await client.count({
    index: 'index',
    body: query,
    pretty: true,
  })
  logResult('countDocuments', res)
}

// Update: Update Document
const updateDocument = async () => {
  const update = {
    doc: {
      field: 'new_value',
    },
  }
  const res = await client.update({
    index: 'index',
    id: 'id',
    body: update,
    refresh: 'true',
  })
  logResult('updateDocument', res)
}

// Update: Upsert Document
const upsertDocument = async () => {
  const upsert = {
    doc: { field: 'new_value' },
    doc_as_upsert: true,
  }
  const res = await client.update({
    index: 'index',
    id: 'id',
    body: upsert,
    refresh: 'true',
  })
  logResult('upsertDocument', res)
}

// Delete: Delete Document
const deleteDocument = async () => {
  const res = await client.delete({
    index: 'index',
    id: 'id',
    refresh: 'true',
  })
  logResult('deleteDocument', res)
}

// Delete: Delete by Query
const deleteByQuery = async () => {
  const query = {
    query: {
      match: {
        field: 'value',
      },
    },
  }
  const res = await client.deleteByQuery({
    index: 'index',
    body: query,
    refresh: 'true',
  })
  logResult('deleteByQuery', res)
}

// Main function to run the examples
const main = async () => {
  await createDocument()
  await bulkIndexDocuments()
  await getDocument()
  await searchDocuments()
  await countDocuments()
  await updateDocument()
  await upsertDocument()
  await deleteDocument()
  await deleteByQuery()
}

main().catch(console.error)

File: ElasticSearch/Go/complete.go
package main

import (
	"context"
	"fmt"
	"github.com/elastic/go-elasticsearch/v7"
	"github.com/elastic/go-elasticsearch/v7/esapi"
	"log"
	"strings"
)

func main() {
	es, err := elasticsearch.NewDefaultClient()
	if err != nil {
		log.Fatalf("Error creating the client: %s", err)
	}

	// Create: Index Document
	createDocument(es)

	// Create: Bulk Index Documents
	bulkIndexDocuments(es)

	// Read: Get Document
	getDocument(es)

	// Read: Search Documents
	searchDocuments(es)

	// Read: Count Documents
	countDocuments(es)

	// Update: Update Document
	updateDocument(es)

	// Update: Upsert Document
	upsertDocument(es)

	// Delete: Delete Document
	deleteDocument(es)

	// Delete: Delete by Query
	deleteByQuery(es)
}

func createDocument(es *elasticsearch.Client) {
	doc := `{"field": "value"}`
	res, err := es.Index(
		"index",
		strings.NewReader(doc),
		es.Index.WithDocumentID("id"),
		es.Index.WithRefresh("true"),
	)
	if err != nil {
		log.Fatalf("Error getting response: %s", err)
	}
	defer res.Body.Close()
	fmt.Println(res)
}

func bulkIndexDocuments(es *elasticsearch.Client) {
	bulk := strings.NewReader(`
    { "index": { "_id": "1" } }
    { "field": "value1" }
    { "index": { "_id": "2" } }
    { "field": "value2" }
    `)
	res, err := es.Bulk(bulk, es.Bulk.WithIndex("index"), es.Bulk.WithRefresh("true"))
	if err != nil {
		log.Fatalf("Error getting response: %s", err)
	}
	defer res.Body.Close()
	fmt.Println(res)
}

func getDocument(es *elasticsearch.Client) {
	res, err := es.Get("index", "id")
	if err != nil {
		log.Fatalf("Error getting document: %s", err)
	}
	defer res.Body.Close()
	fmt.Println(res)
}

func searchDocuments(es *elasticsearch.Client) {
	query := `{
        "query": {
            "match": {
                "field": "value"
            }
        }
    }`
	res, err := es.Search(
		es.Search.WithIndex("index"),
		es.Search.WithBody(strings.NewReader(query)),
		es.Search.WithPretty(),
	)
	if err != nil {
		log.Fatalf("Error searching documents: %s", err)
	}
	defer res.Body.Close()
	fmt.Println(res)
}

func countDocuments(es *elasticsearch.Client) {
	query := `{
        "query": {
            "match_all": {}
        }
    }`
	res, err := es.Count(
		es.Count.WithIndex("index"),
		es.Count.WithBody(strings.NewReader(query)),
		es.Count.WithPretty(),
	)
	if err != nil {
		log.Fatalf("Error counting documents: %s", err)
	}
	defer res.Body.Close()
	fmt.Println(res)
}

func updateDocument(es *elasticsearch.Client) {
	update := `{
        "doc": {
            "field": "new_value"
        }
    }`
	res, err := es.Update(
		"index",
		"id",
		strings.NewReader(update),
		es.Update.WithRefresh("true"),
	)
	if err != nil {
		log.Fatalf("Error updating document: %s", err)
	}
	defer res.Body.Close()
	fmt.Println(res)
}

func upsertDocument(es *elasticsearch.Client) {
	upsert := `{
        "doc": { "field": "new_value" },
        "doc_as_upsert": true
    }`
	res, err := es.Update(
		"index",
		"id",
		strings.NewReader(upsert),
		es.Update.WithRefresh("true"),
	)
	if err != nil {
		log.Fatalf("Error upserting document: %s", err)
	}
	defer res.Body.Close()
	fmt.Println(res)
}

func deleteDocument(es *elasticsearch.Client) {
	res, err := es.Delete(
		"index",
		"id",
		es.Delete.WithRefresh("true"),
	)
	if err != nil {
		log.Fatalf("Error deleting document: %s", err)
	}
	defer res.Body.Close()
	fmt.Println(res)
}

func deleteByQuery(es *elasticsearch.Client) {
	query := `{
        "query": {
            "match": {
                "field": "value"
            }
        }
    }`
	res, err := es.DeleteByQuery(
		[]string{"index"},
		strings.NewReader(query),
		es.DeleteByQuery.WithRefresh("true"),
	)
	if err != nil {
		log.Fatalf("Error deleting by query: %s", err)
	}
	defer res.Body.Close()
	fmt.Println(res)
}
